<a href="https://colab.research.google.com/github/rishabh-0042/American_Express_Hackathon_25-2025/blob/main/Feature_Engineering4.2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
import pandas as pd

folder_path = '/content/drive/MyDrive/Transformed_Train_Data'

# Get a list of all files and directories in the folder
items_in_folder = os.listdir(folder_path)

# Iterate through each item
loaded_data = {}
for item_name in items_in_folder:
    item_path = os.path.join(folder_path, item_name)

    # Skip if it's a directory
    if os.path.isdir(item_path):
        print(f"Skipping directory: {item_name}")
        continue

    try:
        # Try reading as a parquet file first
        if item_name.endswith('.parquet'):
            df = pd.read_parquet(item_path)
            print(f"Shape of {item_name}: {df.shape}")
            loaded_data[item_name] = df
        # Try reading as a CSV with different encodings if parquet fails or it's not a parquet file
        elif item_name.endswith('.csv'):
             try:
                 df = pd.read_csv(item_path)
                 print(f"Shape of {item_name}: {df.shape}")
                 loaded_data[item_name] = df
             except Exception as csv_error:
                 try:
                     df = pd.read_csv(item_path, encoding='latin1')
                     print(f"Shape of {item_name} (latin1 encoding): {df.shape}")
                     loaded_data[item_name] = df
                 except Exception as latin1_error:
                     print(f"Could not load {item_name} with utf-8 or latin1 encoding: {latin1_error}")
        else:
            print(f"Skipping unknown file type: {item_name}")

    except Exception as e:
        print(f"Could not load {item_name}: {e}")

Shape of f226tof309_train_processed.parquet: (770164, 86)
Shape of f310tof314_train_processed.parquet: (770164, 6)
Shape of f315tof319_train_processed.parquet: (770164, 6)
Shape of f320tof324_train_processed.parquet: (770164, 6)
Shape of f325to328_train_processed.parquet: (770164, 5)
Skipping directory: .ipynb_checkpoints
Shape of f329to331_train_processed.parquet: (770164, 2)
Shape of f332to334_train_processed.parquet: (770164, 4)
Shape of train_data_f336to340_imputed.parquet: (770164, 372)
Shape of f341to343_train_processed.parquet: (770164, 2)
Shape of f344to346_train_processed.parquet: (770164, 4)
Shape of f351to357_train_processed.parquet: (770164, 7)
Shape of f354_train_encoded.parquet: (770164, 3)
Shape of f359to366_train_processed.parquet: (770164, 9)
Shape of f347_348_358_train_processed.parquet: (770164, 4)
Shape of f349_350_train_processed.parquet: (770164, 9)


In [ ]:
for file_name, df in loaded_data.items():
    print(f"--- Head of {file_name} ---")
    display(df.head())
    print(f"--- Columns of {file_name} ---")
    print(df.columns)
    print("\n")

--- Head of f226tof309_train_processed.parquet ---


id2  f226  f227  f228  f229  f230  f231  f232  f233  f234  ...  f301  \
0  1366776     0     0     0     0     0     1     0     0     0  ...     0   
1  1366776     0     0     0     0     0     1     0     0     0  ...     0   
2  1366776     0     0     0     0     0     1     0     0     0  ...     0   
3  1366776     0     0     0     0     0     1     0     0     0  ...     0   
4  1366776     0     0     0     0     0     1     0     0     0  ...     0   

   f302  f303  f304  f305  f306  f307  f308  f309  missing_offer_meta  
0     0     0     0     0     0     0     0     0                   0  
1     0     0     0     0     0     0     0     0                   0  
2     0     0     0     0     0     0     0     0                   0  
3     0     0     0     0     0     0     0     0                   0  
4     0     0     0     0     0     0     0     0                   0  

[5 rows x 86 columns]

--- Columns of f226tof309_train_processed.parquet ---
Index(['id2', 'f226', 'f227', 'f228', 'f229', 'f230', 'f231', 'f232', 'f233',
       'f234', 'f235', 'f236', 'f237', 'f238', 'f239', 'f240', 'f241', 'f242',
       'f243', 'f244', 'f245', 'f246', 'f247', 'f248', 'f249', 'f250', 'f251',
       'f252', 'f253', 'f254', 'f255', 'f256', 'f257', 'f258', 'f259', 'f260',
       'f261', 'f262', 'f263', 'f264', 'f265', 'f266', 'f267', 'f268', 'f269',
       'f270', 'f271', 'f272', 'f273', 'f274', 'f275', 'f276', 'f277', 'f278',
       'f279', 'f280', 'f281', 'f282', 'f283', 'f284', 'f285', 'f286', 'f287',
       'f288', 'f289', 'f290', 'f291', 'f292', 'f293', 'f294', 'f295', 'f296',
       'f297', 'f298', 'f299', 'f300', 'f301', 'f302', 'f303', 'f304', 'f305',
       'f306', 'f307', 'f308', 'f309', 'missing_offer_meta'],
      dtype='object')


--- Head of f310tof314_train_processed.parquet ---


f310      f311      f312      f313      f314      id2
0  0.039433  0.037347  0.043935  0.048597  0.053069  1366776
1  0.039433  0.037347  0.043935  0.048597  0.053069  1366776
2  0.039433  0.037347  0.043935  0.048597  0.053069  1366776
3  0.039433  0.037347  0.043935  0.048597  0.053069  1366776
4  0.039433  0.037347  0.043935  0.048597  0.053069  1366776

--- Columns of f310tof314_train_processed.parquet ---
Index(['f310', 'f311', 'f312', 'f313', 'f314', 'id2'], dtype='object')


--- Head of f315tof319_train_processed.parquet ---


f315      f316      f317      f318     f319      id2
0  5.442418  6.477557  7.347283  8.109651  9.09484  1366776
1  5.442418  6.477557  7.347283  8.109651  9.09484  1366776
2  5.442418  6.477557  7.347283  8.109651  9.09484  1366776
3  5.442418  6.477557  7.347283  8.109651  9.09484  1366776
4  5.442418  6.477557  7.347283  8.109651  9.09484  1366776

--- Columns of f315tof319_train_processed.parquet ---
Index(['f315', 'f316', 'f317', 'f318', 'f319', 'id2'], dtype='object')


--- Head of f320tof324_train_processed.parquet ---


id2      f320      f321       f322       f323       f324
0  1366776  8.484318  9.504389  10.346771  11.029478  11.911648
1  1366776  8.484318  9.504389  10.346771  11.029478  11.911648
2  1366776  8.484318  9.504389  10.346771  11.029478  11.911648
3  1366776  8.484318  9.504389  10.346771  11.029478  11.911648
4  1366776  8.484318  9.504389  10.346771  11.029478  11.911648

--- Columns of f320tof324_train_processed.parquet ---
Index(['id2', 'f320', 'f321', 'f322', 'f323', 'f324'], dtype='object')


--- Head of f325to328_train_processed.parquet ---


id2      f325      f326      f327      f328
0  1366776  0.231303  0.191361  0.190917  0.139014
1  1366776  0.231303  0.191361  0.190917  0.139014
2  1366776  0.231303  0.191361  0.190917  0.139014
3  1366776  0.231303  0.191361  0.190917  0.139014
4  1366776  0.231303  0.191361  0.190917  0.139014

--- Columns of f325to328_train_processed.parquet ---
Index(['id2', 'f325', 'f326', 'f327', 'f328'], dtype='object')


--- Head of f329to331_train_processed.parquet ---


id2  f329_to_f331_median
0  1366776             0.288756
1  1366776             0.288756
2  1366776             0.288756
3  1366776             0.288756
4  1366776             0.288756

--- Columns of f329to331_train_processed.parquet ---
Index(['id2', 'f329_to_f331_median'], dtype='object')


--- Head of f332to334_train_processed.parquet ---


id2  f332  f333  f334
0  1366776   1.0   0.0   0.0
1  1366776   1.0   0.0   0.0
2  1366776   1.0   0.0   0.0
3  1366776   1.0   0.0   0.0
4  1366776   1.0   0.0   0.0

--- Columns of f332to334_train_processed.parquet ---
Index(['id2', 'f332', 'f333', 'f334'], dtype='object')


--- Head of train_data_f336to340_imputed.parquet ---


id1      id2        id3  \
0  1366776_189706075_16-23_2023-11-02 22:22:00.042  1366776  189706075   
1      1366776_89227_16-23_2023-11-01 23:51:24.999  1366776      89227   
2      1366776_35046_16-23_2023-11-01 00:30:59.797  1366776      35046   
3    1366776_6275451_16-23_2023-11-02 22:21:32.261  1366776    6275451   
4      1366776_78053_16-23_2023-11-02 22:21:34.799  1366776      78053   

                       id4         id5  y   f1    f2    f3    f4  ...  f357  \
0  2023-11-02 22:22:00.042  2023-11-02  0  1.0  None  None  None  ...  None   
1  2023-11-01 23:51:24.999  2023-11-01  0  1.0  None  None  None  ...  None   
2  2023-11-01 00:30:59.797  2023-11-01  0  1.0  None  None  None  ...  None   
3  2023-11-02 22:21:32.261  2023-11-02  0  1.0  None  None  None  ...  None   
4  2023-11-02 22:21:34.799  2023-11-02  0  1.0  None  None  None  ...  None   

      f358 f359  f360   f361 f362               f363    f364 f365  \
0  -9999.0  0.0  None   28.0  0.0                0.0   337.0  0.0   
1     None  0.0  None   87.0  0.0                0.0  1010.0  2.0   
2     None  0.0  None   23.0  0.0                0.0  1010.0  2.0   
3  -9999.0  0.0  None  277.0  1.0  0.003610108303249   337.0  0.0   
4  -9999.0  0.0  None  359.0  0.0                0.0   337.0  0.0   

                 f366  
0                 0.0  
1  0.0019801980198019  
2  0.0019801980198019  
3                 0.0  
4                 0.0  

[5 rows x 372 columns]

--- Columns of train_data_f336to340_imputed.parquet ---
Index(['id1', 'id2', 'id3', 'id4', 'id5', 'y', 'f1', 'f2', 'f3', 'f4',
       ...
       'f357', 'f358', 'f359', 'f360', 'f361', 'f362', 'f363', 'f364', 'f365',
       'f366'],
      dtype='object', length=372)


--- Head of f341to343_train_processed.parquet ---


id2  f341_to_f343_median
0  1366776             0.606163
1  1366776             0.565762
2  1366776             0.589495
3  1366776             0.596130
4  1366776             0.629771

--- Columns of f341to343_train_processed.parquet ---
Index(['id2', 'f341_to_f343_median'], dtype='object')


--- Head of f344to346_train_processed.parquet ---


id2       f344       f345       f346
0  1366776  12.406066  14.143996  14.143996
1  1366776  12.711250  14.608535  16.355460
2  1366776  12.399041  14.095842  14.095842
3  1366776  12.395473  14.155324  14.155324
4  1366776  12.398930  14.157046  14.157046

--- Columns of f344to346_train_processed.parquet ---
Index(['id2', 'f344', 'f345', 'f346'], dtype='object')


--- Head of f351to357_train_processed.parquet ---


id3  f351  f352  f353      f355       f356      f357
0  189706075   1.0   3.0   0.0  2.798083 -11.038023 -8.608331
1      89227   1.0   3.0   0.0  2.418701  -9.879069 -7.725909
2      35046   1.0   3.0   0.0  2.602570  -8.880616 -6.965691
3    6275451   1.0   3.0   0.0  2.684746  -8.919926 -6.995621
4      78053   1.0   3.0   0.0  2.402651  -9.441466 -7.392720

--- Columns of f351to357_train_processed.parquet ---
Index(['id3', 'f351', 'f352', 'f353', 'f355', 'f356', 'f357'], dtype='object')


--- Head of f354_train_encoded.parquet ---


f354_Phase_1  f354_Rest  f354_nan
0          True      False     False
1         False      False      True
2         False      False      True
3          True      False     False
4          True      False     False

--- Columns of f354_train_encoded.parquet ---
Index(['f354_Phase_1', 'f354_Rest', 'f354_nan'], dtype='object')


--- Head of f359to366_train_processed.parquet ---


id2  f359  f360      f361      f362     f363      f364      f365  \
0  1366776   0.0   1.0  0.315794  0.000000  0.00000  0.289499  0.000000   
1  1366776   0.0   1.0  0.446881  0.000000  0.00000  0.458955  0.166323   
2  1366776   0.0   1.0  0.293447  0.000000  0.00000  0.458955  0.166323   
3  1366776   0.0   1.0  0.582720  0.133766  0.00361  0.289499  0.000000   
4  1366776   0.0   1.0  0.613245  0.000000  0.00000  0.289499  0.000000   

      f366  
0  0.00000  
1  0.00198  
2  0.00198  
3  0.00000  
4  0.00000

--- Columns of f359to366_train_processed.parquet ---
Index(['id2', 'f359', 'f360', 'f361', 'f362', 'f363', 'f364', 'f365', 'f366'], dtype='object')


--- Head of f347_348_358_train_processed.parquet ---


id3      f347      f348      f358
0  189706075  0.267042  0.885361  0.700264
1      89227  0.229358  0.895701  0.226280
2      35046  0.217127  0.887383 -0.182063
3    6275451  0.217654  0.888180 -0.165986
4      78053  0.211024  0.892666  0.047311

--- Columns of f347_348_358_train_processed.parquet ---
Index(['id3', 'f347', 'f348', 'f358'], dtype='object')


--- Head of f349_350_train_processed.parquet ---


id4  f349_1  f349_2  f349_3  f349_4  f349_5  f349_6  \
0  2023-11-02 22:22:00.042     0.0     0.0     0.0     0.0     1.0     0.0   
1  2023-11-01 23:51:24.999     0.0     0.0     0.0     1.0     0.0     0.0   
2  2023-11-01 00:30:59.797     0.0     0.0     0.0     1.0     0.0     0.0   
3  2023-11-02 22:21:32.261     0.0     0.0     0.0     0.0     1.0     0.0   
4  2023-11-02 22:21:34.799     0.0     0.0     0.0     0.0     1.0     0.0   

   f349_7  f350_scaled  
0     0.0     0.931284  
1     0.0     0.993991  
2     0.0     0.021211  
3     0.0     0.931284  
4     0.0     0.931284

--- Columns of f349_350_train_processed.parquet ---
Index(['id4', 'f349_1', 'f349_2', 'f349_3', 'f349_4', 'f349_5', 'f349_6',
       'f349_7', 'f350_scaled'],
      dtype='object')




**Reasoning**:
The subtask is to authenticate and mount Google Drive. The provided notebook already contains a code cell for this purpose. I will execute this cell to complete the subtask.



In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


**Reasoning**:
Load the f335_train_processed.parquet file into a pandas DataFrame.



In [2]:
import pandas as pd

file_path_1 = '/content/drive/MyDrive/Transformed_Test_Data/f142to174_test.parquet'
file_path_2 = '/content/drive/MyDrive/Transformed_Test_Data/f174to309_test.parquet'

f142to174_test = pd.read_parquet(file_path_1)
f174to309_test = pd.read_parquet(file_path_2)

print("Loaded f142to174_test.parquet:")
display(f142to174_test.head())

print("\nLoaded f174to309_test.parquet:")
display(f174to309_test.head())

Loaded f142to174_test.parquet:


id2      f142      f143      f146      f147      f148      f149  \
0  1362907  0.065000  0.050000  0.347449  0.000000  0.338874  0.000000   
1  1082599  0.143678  0.063218  0.674370  0.331294  0.721192  0.275882   
2  1888466  0.038869  0.067138  0.000000  0.334446  0.461679  0.000000   
3  1888971  0.065000  0.050000  0.000000  0.000000  0.000000  0.000000   
4  1256369  0.000000  0.200000  0.000000  0.115757  0.491892  0.000000   

       f150      f151  f152  ...  f165      f166      f167      f168  \
0  0.000000  0.000000   0.0  ...   0.0  0.309938  0.296360  0.035184   
1  0.003607  0.002945   0.0  ...   0.0  0.305589  0.535226  0.035203   
2  0.039263  0.000579   0.0  ...   0.0  0.420284  0.000000  0.001543   
3  0.000846  0.000579   0.0  ...   0.0  0.488307  0.193436  0.040250   
4  0.002489  0.000579   0.0  ...   0.0  0.366250  0.000000  0.000139   

       f169      f170  f171      f172      f173  hours  
0  0.529209  0.444093   0.0  0.000000  0.326724     18  
1  0.522438  0.501890   0.0  0.391097  0.551326      6  
2  0.501892  0.000000   0.0  0.502093  0.380913     10  
3  0.533375  0.545321   0.0  0.000000  0.507293     12  
4  0.436171  0.303526   0.0  0.322769  0.000000      6  

[5 rows x 32 columns]


Loaded f174to309_test.parquet:


id2  f174  f175  f176      f177  f178      f179      f180      f181  \
0  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
1  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
2  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
3  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
4  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   

   f182  ...  f301  f302  f303  f304  f305  f306  f307  f308  f309  \
0   0.0  ...     0     0     0     0     0     0     0     0     0   
1   0.0  ...     0     0     0     0     0     0     0     0     0   
2   0.0  ...     0     0     0     0     0     0     0     0     0   
3   0.0  ...     0     0     0     0     0     0     0     0     0   
4   0.0  ...     0     0     0     0     0     0     0     0     0   

   missing_offer_meta  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 138 columns]

In [3]:
# Drop 'id2' from f174to309_test
f174to309_test_modified = f174to309_test.drop('id2', axis=1)
print("Shape of f174to309_test after dropping 'id2':", f174to309_test_modified.shape)

# Drop 'hours' from f142to174_test
f142to174_test_modified = f142to174_test.drop('hours', axis=1)
print("Shape of f142to174_test after dropping 'hours':", f142to174_test_modified.shape)

# Merge the modified dataframes based on index
f142to309 = pd.merge(f142to174_test_modified, f174to309_test_modified, left_index=True, right_index=True)

print("Shape of the merged dataframe f142to309:", f142to309.shape)

# Rename the merged dataframe
f142to309_test = f142to309

print("\nHead of the renamed and merged dataframe f142to309_test:")
display(f142to309_test.head())

# Confirm the original f142to174_test is still available
print("\nShape of the original f142to174_test:", f142to174_test.shape)

Shape of f174to309_test after dropping 'id2': (369301, 137)
Shape of f142to174_test after dropping 'hours': (369301, 31)
Shape of the merged dataframe f142to309: (369301, 168)

Head of the renamed and merged dataframe f142to309_test:


id2      f142      f143      f146      f147      f148      f149  \
0  1362907  0.065000  0.050000  0.347449  0.000000  0.338874  0.000000   
1  1082599  0.143678  0.063218  0.674370  0.331294  0.721192  0.275882   
2  1888466  0.038869  0.067138  0.000000  0.334446  0.461679  0.000000   
3  1888971  0.065000  0.050000  0.000000  0.000000  0.000000  0.000000   
4  1256369  0.000000  0.200000  0.000000  0.115757  0.491892  0.000000   

       f150      f151  f152  ...  f301  f302  f303  f304  f305  f306  f307  \
0  0.000000  0.000000   0.0  ...     0     0     0     0     0     0     0   
1  0.003607  0.002945   0.0  ...     0     0     0     0     0     0     0   
2  0.039263  0.000579   0.0  ...     0     0     0     0     0     0     0   
3  0.000846  0.000579   0.0  ...     0     0     0     0     0     0     0   
4  0.002489  0.000579   0.0  ...     0     0     0     0     0     0     0   

   f308  f309  missing_offer_meta  
0     0     0                   0  
1     0     0                   0  
2     0     0                   0  
3     0     0                   0  
4     0     0                   0  

[5 rows x 168 columns]


Shape of the original f142to174_test: (369301, 32)


In [4]:
output_path = '/content/drive/MyDrive/Transformed_Test_Data/f142to309_test.parquet'
f142to309_test.to_parquet(output_path, index=False)

print(f"DataFrame f142to309_test saved to {output_path}")

DataFrame f142to309_test saved to /content/drive/MyDrive/Transformed_Test_Data/f142to309_test.parquet


In [5]:
import pandas as pd

file_path_train_1 = '/content/drive/MyDrive/Transformed_Train_Data/f142to174_train.parquet'
file_path_train_2 = '/content/drive/MyDrive/Transformed_Train_Data/f174to309_train.parquet'

f142to174_train = pd.read_parquet(file_path_train_1)
f174to309_train = pd.read_parquet(file_path_train_2)

print("Loaded f142to174_train.parquet:")
display(f142to174_train.head())

print("\nLoaded f174to309_train.parquet:")
display(f174to309_train.head())

Loaded f142to174_train.parquet:


id2      f142      f143      f146      f147      f148      f149  \
0  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   
1  1366776  0.181818  0.272727  0.699950  0.173474  0.678726  0.186631   
2  1366776  0.142857  0.285714  0.591029  0.079303  0.573107  0.085318   
3  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   
4  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   

       f150      f151  f152  ...  f165      f166  f167      f168      f169  \
0  0.000902  0.000965   0.0  ...   0.0  0.313445   0.0  0.003835  0.487751   
1  0.000902  0.000878   0.0  ...   0.0  0.313445   0.0  0.002672  0.487751   
2  0.000626  0.000609   0.0  ...   0.0  0.313445   0.0  0.002483  0.487751   
3  0.000902  0.000965   0.0  ...   0.0  0.313445   0.0  0.003835  0.487751   
4  0.000902  0.000965   0.0  ...   0.0  0.313445   0.0  0.003835  0.487751   

       f170  f171      f172      f173  hours  
0  0.321342   0.0  0.316842  0.379028     22  
1  0.321342   0.0  0.316842  0.379028     23  
2  0.321342   0.0  0.316842  0.379028      0  
3  0.321342   0.0  0.316842  0.379028     22  
4  0.321342   0.0  0.316842  0.379028     22  

[5 rows x 32 columns]


Loaded f174to309_train.parquet:


id2  f174  f175  f176      f177  f178      f179      f180      f181  \
0  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
1  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
2  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
3  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   
4  1366776   0.0   0.0   0.0  0.099722   0.0  0.388228  0.423617  0.144152   

   f182  ...  f301  f302  f303  f304  f305  f306  f307  f308  f309  \
0   0.0  ...     0     0     0     0     0     0     0     0     0   
1   0.0  ...     0     0     0     0     0     0     0     0     0   
2   0.0  ...     0     0     0     0     0     0     0     0     0   
3   0.0  ...     0     0     0     0     0     0     0     0     0   
4   0.0  ...     0     0     0     0     0     0     0     0     0   

   missing_offer_meta  
0                   0  
1                   0  
2                   0  
3                   0  
4                   0  

[5 rows x 138 columns]

In [6]:
# 1. Show initial shapes
print("Initial shape of f142to174_train:", f142to174_train.shape)
print("Initial shape of f174to309_train:", f174to309_train.shape)

# 2. Drop columns
f174to309_train_modified = f174to309_train.drop('id2', axis=1)
print("\nShape of f174to309_train after dropping 'id2':", f174to309_train_modified.shape)

f142to174_train_modified = f142to174_train.drop('hours', axis=1)
print("Shape of f142to174_train after dropping 'hours':", f142to174_train_modified.shape)

# 3. Merge dataframes based on index
f142to309_train_merged = pd.merge(f142to174_train_modified, f174to309_train_modified, left_index=True, right_index=True)
print("\nShape of the merged dataframe:", f142to309_train_merged.shape)

# 4. Rename merged dataframe
f142to309_train = f142to309_train_merged

# 5. Verify changes
print("\nHead of the renamed and merged dataframe f142to309_train:")
display(f142to309_train.head())

print("\nShape of the original f142to174_train:", f142to174_train.shape)

Initial shape of f142to174_train: (770164, 32)
Initial shape of f174to309_train: (770164, 138)

Shape of f174to309_train after dropping 'id2': (770164, 137)
Shape of f142to174_train after dropping 'hours': (770164, 31)

Shape of the merged dataframe: (770164, 168)

Head of the renamed and merged dataframe f142to309_train:


id2      f142      f143      f146      f147      f148      f149  \
0  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   
1  1366776  0.181818  0.272727  0.699950  0.173474  0.678726  0.186631   
2  1366776  0.142857  0.285714  0.591029  0.079303  0.573107  0.085318   
3  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   
4  1366776  0.181818  0.272727  0.653985  0.167984  0.671533  0.158738   

       f150      f151  f152  ...  f301  f302  f303  f304  f305  f306  f307  \
0  0.000902  0.000965   0.0  ...     0     0     0     0     0     0     0   
1  0.000902  0.000878   0.0  ...     0     0     0     0     0     0     0   
2  0.000626  0.000609   0.0  ...     0     0     0     0     0     0     0   
3  0.000902  0.000965   0.0  ...     0     0     0     0     0     0     0   
4  0.000902  0.000965   0.0  ...     0     0     0     0     0     0     0   

   f308  f309  missing_offer_meta  
0     0     0                   0  
1     0     0                   0  
2     0     0                   0  
3     0     0                   0  
4     0     0                   0  

[5 rows x 168 columns]


Shape of the original f142to174_train: (770164, 32)


In [7]:
output_path = '/content/drive/MyDrive/Transformed_Train_Data/f142to309_train.parquet'
f142to309_train.to_parquet(output_path, index=False)

print(f"DataFrame f142to309_train saved to {output_path}")

DataFrame f142to309_train saved to /content/drive/MyDrive/Transformed_Train_Data/f142to309_train.parquet


In [8]:
import os

folder_path = '/content/drive/MyDrive/AMEXChallenge'
items_in_folder = os.listdir(folder_path)
print(items_in_folder)

['ps.xlsx', 'train_data.parquet', 'test_data.parquet', 'add_trans.parquet', 'add_event.parquet', 'offer_metadata.parquet', 'data_dictionary.csv', '685404e30cfdb_submission_template.csv', 'interest_imputed.parquet', 'f13tof21_transformed.parquet', '.ipynb_checkpoints']


In [9]:
import pandas as pd

train_data_path = '/content/drive/MyDrive/AMEXChallenge/train_data.parquet'
test_data_path = '/content/drive/MyDrive/AMEXChallenge/test_data.parquet'

train_data = pd.read_parquet(train_data_path)
test_data = pd.read_parquet(test_data_path)

print("Loaded train_data.parquet:")
display(train_data.head())

print("\nLoaded test_data.parquet:")
display(test_data.head())

Loaded train_data.parquet:


id1      id2        id3  \
0  1366776_189706075_16-23_2023-11-02 22:22:00.042  1366776  189706075   
1      1366776_89227_16-23_2023-11-01 23:51:24.999  1366776      89227   
2      1366776_35046_16-23_2023-11-01 00:30:59.797  1366776      35046   
3    1366776_6275451_16-23_2023-11-02 22:21:32.261  1366776    6275451   
4      1366776_78053_16-23_2023-11-02 22:21:34.799  1366776      78053   

                       id4         id5  y   f1    f2    f3    f4  ...  f357  \
0  2023-11-02 22:22:00.042  2023-11-02  0  1.0  None  None  None  ...  None   
1  2023-11-01 23:51:24.999  2023-11-01  0  1.0  None  None  None  ...  None   
2  2023-11-01 00:30:59.797  2023-11-01  0  1.0  None  None  None  ...  None   
3  2023-11-02 22:21:32.261  2023-11-02  0  1.0  None  None  None  ...  None   
4  2023-11-02 22:21:34.799  2023-11-02  0  1.0  None  None  None  ...  None   

      f358 f359  f360   f361 f362               f363    f364 f365  \
0  -9999.0  0.0  None   28.0  0.0                0.0   337.0  0.0   
1     None  0.0  None   87.0  0.0                0.0  1010.0  2.0   
2     None  0.0  None   23.0  0.0                0.0  1010.0  2.0   
3  -9999.0  0.0  None  277.0  1.0  0.003610108303249   337.0  0.0   
4  -9999.0  0.0  None  359.0  0.0                0.0   337.0  0.0   

                 f366  
0                 0.0  
1  0.0019801980198019  
2  0.0019801980198019  
3                 0.0  
4                 0.0  

[5 rows x 372 columns]


Loaded test_data.parquet:


id1      id2     id3  \
46756    1362907_91950_16-23_2023-11-04 18:56:26.000794  1362907   91950   
57819       1082599_88356_16-23_2023-11-04 06:08:53.373  1082599   88356   
15390   1888466_958700_16-23_2023-11-05 10:07:28.000725  1888466  958700   
145730     1888971_795739_16-23_2023-11-04 12:25:28.244  1888971  795739   
146085      1256369_82296_16-23_2023-11-05 06:45:26.657  1256369   82296   

                               id4         id5    f1    f2    f3    f4    f5  \
46756   2023-11-04 18:56:26.000794  2023-11-04  None  None  None  None  None   
57819      2023-11-04 06:08:53.373  2023-11-04  None   9.0  None  None  None   
15390   2023-11-05 10:07:28.000725  2023-11-05  None  None  None  None  22.0   
145730     2023-11-04 12:25:28.244  2023-11-04  None  None  None  None  None   
146085     2023-11-05 06:45:26.657  2023-11-05  None  None  None  None  None   

        ...                 f357                f358 f359  f360   f361  f362  \
46756   ...   0.0027807272790246  0.0465999838670646  0.0  None    1.0   0.0   
57819   ...   0.0014287946777398  0.0603093514970909  0.0  None  195.0  13.0   
15390   ...  -0.0174961119751166  0.0734836702954899  0.0  None  155.0  67.0   
145730  ...   0.0013159476327096   0.040572039549215  0.0  None   None  None   
146085  ...   0.0020541699799278   0.038243539079214  0.0  None   29.0   2.0   

                      f363    f364   f365                f366  
46756                  0.0    56.0    0.0                 0.0  
57819   0.0666666666666666    None   None                None  
15390    0.432258064516129  1142.0  436.0  0.3817863397548161  
145730                None    None   None                None  
146085  0.0689655172413793   361.0    3.0  0.0083102493074792  

[5 rows x 371 columns]

In [10]:
# Select columns for f130to141_train
f130to141_train_columns = ['id2', 'id3'] + [f'f{i}' for i in range(130, 142)]
f130to141_train = train_data[f130to141_train_columns].copy()

# Select columns for f130to141_test
f130to141_test_columns = ['id2', 'id3'] + [f'f{i}' for i in range(130, 142)]
f130to141_test = test_data[f130to141_test_columns].copy()

print("Head of f130to141_train:")
display(f130to141_train.head())

print("\nHead of f130to141_test:")
display(f130to141_test.head())

Head of f130to141_train:


id2        id3                f130                f131  \
0  1366776  189706075              0.0125  0.0555555555555555   
1  1366776      89227              0.0125  0.0555555555555555   
2  1366776      35046  0.0158730158730158  0.0350877192982456   
3  1366776    6275451              0.0125  0.0555555555555555   
4  1366776      78053              0.0125  0.0555555555555555   

                 f132                f133                f134  f135  f136  \
0  0.0005313496280552  0.0117647058823529  0.0157894736842105  None  None   
1  0.0005313496280552  0.0117647058823529  0.0157894736842105  None  None   
2  0.0006553079947575  0.0070921985815602  0.0130718954248366  None  None   
3  0.0005313496280552  0.0117647058823529  0.0157894736842105  None  None   
4  0.0005313496280552  0.0117647058823529  0.0157894736842105  None  None   

                 f137                f138                f139  \
0   0.004594820384294  0.0070351758793969  1.5311100959342168   
1   0.004594820384294  0.0070351758793969  1.5311100959342168   
2  0.0036082474226804  0.0055452865064695    1.53683654607869   
3   0.004594820384294  0.0070351758793969  1.5311100959342168   
4   0.004594820384294  0.0070351758793969  1.5311100959342168   

                 f140                f141  
0  0.0909090909090909  0.3636363636363636  
1  0.0909090909090909  0.3636363636363636  
2  0.1428571428571428  0.2857142857142857  
3  0.0909090909090909  0.3636363636363636  
4  0.0909090909090909  0.3636363636363636


Head of f130to141_test:


id2     id3                f130                f131  \
46756   1362907   91950                 0.0                 0.0   
57819   1082599   88356  0.0222602739726027  0.0277777777777777   
15390   1888466  958700  0.3333333333333333  0.4545454545454545   
145730  1888971  795739                 0.0                None   
146085  1256369   82296                 0.0  0.1176470588235294   

                      f132                f133                f134  f135  \
46756                  0.0                 0.0                 0.0  None   
57819   0.0276306700437485  0.0806451612903225  0.0317919075144508  None   
15390    0.425569176882662  0.3666666666666666  0.4634146341463415  None   
145730                 0.0                None                None  None   
146085   0.010752688172043                 0.0  0.0344827586206896  None   

        f136                f137                f138                f139  \
46756   None                 0.0                 0.0                None   
57819   None  0.0303664921465968  0.0388535031847133   1.279486053151768   
15390   None  0.4236526946107784  0.3990147783251231  0.9418440703928702   
145730  None                 0.0                None                None   
146085  None  0.0139664804469273  0.0188679245283018   1.350943396226415   

                      f140                f141  
46756                 None                None  
57819   0.0747126436781609  0.0229885057471264  
15390   0.0176678445229682  0.0176678445229682  
145730                None                None  
146085                 0.0                 0.2

In [11]:
import numpy as np

# Create a copy to avoid modifying the original dataframe
df_analysis = f130to141_train.copy()

# Check data types and attempt conversion to numeric
for col in df_analysis.columns:
    # Attempt to convert to numeric, coercing errors to NaN
    df_analysis[col] = pd.to_numeric(df_analysis[col], errors='coerce')

# Calculate null percentages and range for each column
analysis_results = {}
for col in df_analysis.columns:
    null_percentage = df_analysis[col].isnull().mean() * 100
    col_range = None
    if pd.api.types.is_numeric_dtype(df_analysis[col]):
        # Calculate range only for numeric columns, excluding NaN
        if df_analysis[col].notna().any():
            col_range = df_analysis[col].max() - df_analysis[col].min()

    analysis_results[col] = {
        'null_percentage': null_percentage,
        'range': col_range
    }

# Display the results
print("Analysis of f130to141_train:")
for col, results in analysis_results.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['range'] is not None:
        print(f"  Range: {results['range']:.4f}")
    else:
        print("  Range: N/A (column is not numeric or all values are null)")
    print("-" * 20)

Analysis of f130to141_train:
Column: id2
  Null Percentage: 0.00%
  Range: 910525.0000
--------------------
Column: id3
  Null Percentage: 0.00%
  Range: 876664650.0000
--------------------
Column: f130
  Null Percentage: 12.51%
  Range: 1.0000
--------------------
Column: f131
  Null Percentage: 15.39%
  Range: 1.0000
--------------------
Column: f132
  Null Percentage: 4.95%
  Range: 1.0000
--------------------
Column: f133
  Null Percentage: 14.17%
  Range: 1.0000
--------------------
Column: f134
  Null Percentage: 10.58%
  Range: 1.0000
--------------------
Column: f135
  Null Percentage: 100.00%
  Range: N/A (column is not numeric or all values are null)
--------------------
Column: f136
  Null Percentage: 100.00%
  Range: N/A (column is not numeric or all values are null)
--------------------
Column: f137
  Null Percentage: 4.78%
  Range: 1.0000
--------------------
Column: f138
  Null Percentage: 9.74%
  Range: 1.0000
--------------------
Column: f139
  Null Percentage: 29.83%


In [12]:
import numpy as np

# Create a copy to avoid modifying the original dataframe
df_analysis_test = f130to141_test.copy()

# Check data types and attempt conversion to numeric
for col in df_analysis_test.columns:
    # Attempt to convert to numeric, coercing errors to NaN
    df_analysis_test[col] = pd.to_numeric(df_analysis_test[col], errors='coerce')

# Calculate null percentages and range for each column
analysis_results_test = {}
for col in df_analysis_test.columns:
    null_percentage = df_analysis_test[col].isnull().mean() * 100
    col_range = None
    if pd.api.types.is_numeric_dtype(df_analysis_test[col]):
        # Calculate range only for numeric columns, excluding NaN
        if df_analysis_test[col].notna().any():
            col_range = df_analysis_test[col].max() - df_analysis_test[col].min()

    analysis_results_test[col] = {
        'null_percentage': null_percentage,
        'range': col_range
    }

# Display the results
print("Analysis of f130to141_test:")
for col, results in analysis_results_test.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['range'] is not None:
        print(f"  Range: {results['range']:.4f}")
    else:
        print("  Range: N/A (column is not numeric or all values are null)")
    print("-" * 20)

Analysis of f130to141_test:
Column: id2
  Null Percentage: 0.00%
  Range: 910504.0000
--------------------
Column: id3
  Null Percentage: 0.00%
  Range: 876664650.0000
--------------------
Column: f130
  Null Percentage: 12.15%
  Range: 1.0000
--------------------
Column: f131
  Null Percentage: 15.07%
  Range: 1.0000
--------------------
Column: f132
  Null Percentage: 5.10%
  Range: 1.0000
--------------------
Column: f133
  Null Percentage: 13.93%
  Range: 1.0000
--------------------
Column: f134
  Null Percentage: 9.99%
  Range: 1.0000
--------------------
Column: f135
  Null Percentage: 100.00%
  Range: N/A (column is not numeric or all values are null)
--------------------
Column: f136
  Null Percentage: 100.00%
  Range: N/A (column is not numeric or all values are null)
--------------------
Column: f137
  Null Percentage: 4.85%
  Range: 1.0000
--------------------
Column: f138
  Null Percentage: 9.79%
  Range: 1.0000
--------------------
Column: f139
  Null Percentage: 28.83%
  

**Reasoning**:
Drop columns 'f135' and 'f136' from both dataframes and display the head and shape to confirm the changes.



In [13]:
f130to141_train_modified = f130to141_train.drop(['f135', 'f136'], axis=1)
f130to141_test_modified = f130to141_test.drop(['f135', 'f136'], axis=1)

print("Head of f130to141_train after dropping f135 and f136:")
display(f130to141_train_modified.head())
print("\nShape of f130to141_train after dropping f135 and f136:", f130to141_train_modified.shape)

print("\nHead of f130to141_test after dropping f135 and f136:")
display(f130to141_test_modified.head())
print("\nShape of f130to141_test after dropping f135 and f136:", f130to141_test_modified.shape)

Head of f130to141_train after dropping f135 and f136:


id2        id3                f130                f131  \
0  1366776  189706075              0.0125  0.0555555555555555   
1  1366776      89227              0.0125  0.0555555555555555   
2  1366776      35046  0.0158730158730158  0.0350877192982456   
3  1366776    6275451              0.0125  0.0555555555555555   
4  1366776      78053              0.0125  0.0555555555555555   

                 f132                f133                f134  \
0  0.0005313496280552  0.0117647058823529  0.0157894736842105   
1  0.0005313496280552  0.0117647058823529  0.0157894736842105   
2  0.0006553079947575  0.0070921985815602  0.0130718954248366   
3  0.0005313496280552  0.0117647058823529  0.0157894736842105   
4  0.0005313496280552  0.0117647058823529  0.0157894736842105   

                 f137                f138                f139  \
0   0.004594820384294  0.0070351758793969  1.5311100959342168   
1   0.004594820384294  0.0070351758793969  1.5311100959342168   
2  0.0036082474226804  0.0055452865064695    1.53683654607869   
3   0.004594820384294  0.0070351758793969  1.5311100959342168   
4   0.004594820384294  0.0070351758793969  1.5311100959342168   

                 f140                f141  
0  0.0909090909090909  0.3636363636363636  
1  0.0909090909090909  0.3636363636363636  
2  0.1428571428571428  0.2857142857142857  
3  0.0909090909090909  0.3636363636363636  
4  0.0909090909090909  0.3636363636363636


Shape of f130to141_train after dropping f135 and f136: (770164, 12)

Head of f130to141_test after dropping f135 and f136:


id2     id3                f130                f131  \
46756   1362907   91950                 0.0                 0.0   
57819   1082599   88356  0.0222602739726027  0.0277777777777777   
15390   1888466  958700  0.3333333333333333  0.4545454545454545   
145730  1888971  795739                 0.0                None   
146085  1256369   82296                 0.0  0.1176470588235294   

                      f132                f133                f134  \
46756                  0.0                 0.0                 0.0   
57819   0.0276306700437485  0.0806451612903225  0.0317919075144508   
15390    0.425569176882662  0.3666666666666666  0.4634146341463415   
145730                 0.0                None                None   
146085   0.010752688172043                 0.0  0.0344827586206896   

                      f137                f138                f139  \
46756                  0.0                 0.0                None   
57819   0.0303664921465968  0.0388535031847133   1.279486053151768   
15390   0.4236526946107784  0.3990147783251231  0.9418440703928702   
145730                 0.0                None                None   
146085  0.0139664804469273  0.0188679245283018   1.350943396226415   

                      f140                f141  
46756                 None                None  
57819   0.0747126436781609  0.0229885057471264  
15390   0.0176678445229682  0.0176678445229682  
145730                None                None  
146085                 0.0                 0.2


Shape of f130to141_test after dropping f135 and f136: (369301, 12)


**Reasoning**:
Identify and print the numeric columns in the modified dataframes.



In [14]:
import pandas as pd

numeric_cols_train = [col for col in f130to141_train_modified.columns if pd.api.types.is_numeric_dtype(f130to141_train_modified[col])]
numeric_cols_test = [col for col in f130to141_test_modified.columns if pd.api.types.is_numeric_dtype(f130to141_test_modified[col])]

print("Numeric columns in f130to141_train_modified:")
print(numeric_cols_train)

print("\nNumeric columns in f130to141_test_modified:")
print(numeric_cols_test)

Numeric columns in f130to141_train_modified:
[]

Numeric columns in f130to141_test_modified:
[]


**Reasoning**:
The previous attempt to identify numeric columns resulted in empty lists, indicating that the columns are not currently in a numeric format. I need to convert the relevant columns to numeric types before identifying them.



In [15]:
import numpy as np

# Convert columns to numeric, coercing errors to NaN
for col in f130to141_train_modified.columns:
    f130to141_train_modified[col] = pd.to_numeric(f130to141_train_modified[col], errors='coerce')

for col in f130to141_test_modified.columns:
    f130to141_test_modified[col] = pd.to_numeric(f130to141_test_modified[col], errors='coerce')

# Now identify numeric columns again
numeric_cols_train = [col for col in f130to141_train_modified.columns if pd.api.types.is_numeric_dtype(f130to141_train_modified[col])]
numeric_cols_test = [col for col in f130to141_test_modified.columns if pd.api.types.is_numeric_dtype(f130to141_test_modified[col])]

print("Numeric columns in f130to141_train_modified after conversion:")
print(numeric_cols_train)

print("\nNumeric columns in f130to141_test_modified after conversion:")
print(numeric_cols_test)

Numeric columns in f130to141_train_modified after conversion:
['id2', 'id3', 'f130', 'f131', 'f132', 'f133', 'f134', 'f137', 'f138', 'f139', 'f140', 'f141']

Numeric columns in f130to141_test_modified after conversion:
['id2', 'id3', 'f130', 'f131', 'f132', 'f133', 'f134', 'f137', 'f138', 'f139', 'f140', 'f141']


**Reasoning**:
Calculate the minimum and maximum values for each numeric column in both dataframes and store them.



In [16]:
train_numeric_range = {}
for col in numeric_cols_train:
    train_numeric_range[col] = {
        'min': f130to141_train_modified[col].min(),
        'max': f130to141_train_modified[col].max()
    }

test_numeric_range = {}
for col in numeric_cols_test:
    test_numeric_range[col] = {
        'min': f130to141_test_modified[col].min(),
        'max': f130to141_test_modified[col].max()
    }

print("Numeric column ranges for f130to141_train_modified:")
for col, values in train_numeric_range.items():
    print(f"  {col}: Min = {values['min']:.4f}, Max = {values['max']:.4f}")

print("\nNumeric column ranges for f130to141_test_modified:")
for col, values in test_numeric_range.items():
    print(f"  {col}: Min = {values['min']:.4f}, Max = {values['max']:.4f}")

Numeric column ranges for f130to141_train_modified:
  id2: Min = 1000043.0000, Max = 1910568.0000
  id3: Min = 1185.0000, Max = 876665835.0000
  f130: Min = 0.0000, Max = 1.0000
  f131: Min = 0.0000, Max = 1.0000
  f132: Min = 0.0000, Max = 1.0000
  f133: Min = 0.0000, Max = 1.0000
  f134: Min = 0.0000, Max = 1.0000
  f137: Min = 0.0000, Max = 1.0000
  f138: Min = 0.0000, Max = 1.0000
  f139: Min = 0.0000, Max = 310.1153
  f140: Min = 0.0000, Max = 1.0000
  f141: Min = 0.0000, Max = 1.0000

Numeric column ranges for f130to141_test_modified:
  id2: Min = 1000061.0000, Max = 1910565.0000
  id3: Min = 1185.0000, Max = 876665835.0000
  f130: Min = 0.0000, Max = 1.0000
  f131: Min = 0.0000, Max = 1.0000
  f132: Min = 0.0000, Max = 1.0000
  f133: Min = 0.0000, Max = 1.0000
  f134: Min = 0.0000, Max = 1.0000
  f137: Min = 0.0000, Max = 1.0000
  f138: Min = 0.0000, Max = 1.0000
  f139: Min = 0.0000, Max = 300.0000
  f140: Min = 0.0000, Max = 1.0000
  f141: Min = 0.0000, Max = 1.0000


**Reasoning**:
The subtask is to present the minimum and maximum values for each numeric column in both train and test dataframes, which are stored in the `train_numeric_range` and `test_numeric_range` dictionaries. I will print these dictionaries to display the results.



In [17]:
print("Minimum and Maximum values for numeric columns in f130to141_train_modified:")
for col, values in train_numeric_range.items():
    print(f"  {col}: Min = {values['min']:.4f}, Max = {values['max']:.4f}")

print("\nMinimum and Maximum values for numeric columns in f130to141_test_modified:")
for col, values in test_numeric_range.items():
    print(f"  {col}: Min = {values['min']:.4f}, Max = {values['max']:.4f}")

Minimum and Maximum values for numeric columns in f130to141_train_modified:
  id2: Min = 1000043.0000, Max = 1910568.0000
  id3: Min = 1185.0000, Max = 876665835.0000
  f130: Min = 0.0000, Max = 1.0000
  f131: Min = 0.0000, Max = 1.0000
  f132: Min = 0.0000, Max = 1.0000
  f133: Min = 0.0000, Max = 1.0000
  f134: Min = 0.0000, Max = 1.0000
  f137: Min = 0.0000, Max = 1.0000
  f138: Min = 0.0000, Max = 1.0000
  f139: Min = 0.0000, Max = 310.1153
  f140: Min = 0.0000, Max = 1.0000
  f141: Min = 0.0000, Max = 1.0000

Minimum and Maximum values for numeric columns in f130to141_test_modified:
  id2: Min = 1000061.0000, Max = 1910565.0000
  id3: Min = 1185.0000, Max = 876665835.0000
  f130: Min = 0.0000, Max = 1.0000
  f131: Min = 0.0000, Max = 1.0000
  f132: Min = 0.0000, Max = 1.0000
  f133: Min = 0.0000, Max = 1.0000
  f134: Min = 0.0000, Max = 1.0000
  f137: Min = 0.0000, Max = 1.0000
  f138: Min = 0.0000, Max = 1.0000
  f139: Min = 0.0000, Max = 300.0000
  f140: Min = 0.0000, Max = 1.00

## Summary:

### Data Analysis Key Findings

*   The columns 'f135' and 'f136' were successfully dropped from both the training and testing dataframes, reducing the number of columns from 14 to 12.
*   After converting the relevant columns to a numeric data type, the numeric columns were identified in both the modified training and testing dataframes.
*   The minimum and maximum values for each numeric column were calculated and displayed for both the training and testing datasets. For example, in the training data, `id2` ranges from 1000043.0000 to 1910568.0000, while `f139` ranges from 0 to 310.1153. In the test data, `f139` ranges from 0 to 300.0000.

### Insights or Next Steps

*   The calculated ranges provide an initial understanding of the distribution and scale of the numeric features, which can be useful for feature scaling or outlier detection in subsequent steps.
*   Further analysis could involve comparing the ranges of corresponding columns between the training and testing sets to identify potential data shifts or inconsistencies.


**Reasoning**:
Define the lists of column names for the two imputation strategies as instructed.



In [18]:
id3_grouped_cols = ['f130', 'f131', 'f132', 'f133', 'f134', 'f137', 'f138', 'f139']
id2_grouped_cols = ['f140', 'f141']

print("Columns to be grouped by 'id3':")
print(id3_grouped_cols)

print("\nColumns to be grouped by 'id2':")
print(id2_grouped_cols)

Columns to be grouped by 'id3':
['f130', 'f131', 'f132', 'f133', 'f134', 'f137', 'f138', 'f139']

Columns to be grouped by 'id2':
['f140', 'f141']


**Reasoning**:
Impute missing values in the specified columns of the training data by grouping by 'id3' and using the median, with a fallback to the global median.



In [19]:
# Create a copy to avoid modifying the original dataframe
f130to141_train_imputed_id3 = f130to141_train_modified.copy()

# Impute missing values in id3_grouped_cols
for col in id3_grouped_cols:
    # Calculate the global median for the current column
    global_median = f130to141_train_imputed_id3[col].median()

    # Group by 'id3' and fill missing values within each group with the group median
    # Use transform to return a Series with the same index as the original DataFrame
    f130to141_train_imputed_id3[col] = f130to141_train_imputed_id3.groupby('id3')[col].transform(
        lambda x: x.fillna(x.median())
    )

    # Fill any remaining missing values with the global median
    f130to141_train_imputed_id3[col] = f130to141_train_imputed_id3[col].fillna(global_median)

# Display the head and check for null values in the imputed columns
print("Head of f130to141_train_imputed_id3 after imputation:")
display(f130to141_train_imputed_id3.head())

print("\nChecking for null values in imputed columns (id3 grouped):")
print(f130to141_train_imputed_id3[id3_grouped_cols].isnull().sum())

/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


Head of f130to141_train_imputed_id3 after imputation:


/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


id2        id3      f130      f131      f132      f133      f134  \
0  1366776  189706075  0.012500  0.055556  0.000531  0.011765  0.015789   
1  1366776      89227  0.012500  0.055556  0.000531  0.011765  0.015789   
2  1366776      35046  0.015873  0.035088  0.000655  0.007092  0.013072   
3  1366776    6275451  0.012500  0.055556  0.000531  0.011765  0.015789   
4  1366776      78053  0.012500  0.055556  0.000531  0.011765  0.015789   

       f137      f138      f139      f140      f141  
0  0.004595  0.007035  1.531110  0.090909  0.363636  
1  0.004595  0.007035  1.531110  0.090909  0.363636  
2  0.003608  0.005545  1.536837  0.142857  0.285714  
3  0.004595  0.007035  1.531110  0.090909  0.363636  
4  0.004595  0.007035  1.531110  0.090909  0.363636


Checking for null values in imputed columns (id3 grouped):
f130    0
f131    0
f132    0
f133    0
f134    0
f137    0
f138    0
f139    0
dtype: int64


In [20]:
output_path = '/content/drive/MyDrive/Transformed_Train_Data/f130to141_train_imputed_id3.parquet'
f130to141_train_imputed_id3.to_parquet(output_path, index=False)

print(f"DataFrame f130to141_train_imputed_id3 saved to {output_path}")

DataFrame f130to141_train_imputed_id3 saved to /content/drive/MyDrive/Transformed_Train_Data/f130to141_train_imputed_id3.parquet


**Reasoning**:
Impute missing values in columns f140-f141 of dataframe f130to141_train_imputed_id3 by grouping by 'id2' and using the median of non-null values within each group, falling back to the global median if a group has no non-null values.



In [21]:
# Create a copy of the DataFrame
f130to141_train_imputed_id2_id3 = f130to141_train_imputed_id3.copy()

# Define the columns to be imputed in this subtask
id2_grouped_cols = ['f140', 'f141']

# Iterate through each column and perform grouped median imputation with global median fallback
for col in id2_grouped_cols:
    # Calculate the global median for the column
    global_median = f130to141_train_imputed_id2_id3[col].median()

    # Impute missing values by grouping by 'id2' and using the median of non-null values,
    # falling back to the global median if a group has no non-null values.
    f130to141_train_imputed_id2_id3[col] = f130to141_train_imputed_id2_id3.groupby('id2')[col].transform(
        lambda x: x.fillna(x.median() if pd.notna(x.median()) else global_median)
    )

# Display the head of the imputed DataFrame
print("Head of f130to141_train_imputed_id2_id3 after imputation:")
display(f130to141_train_imputed_id2_id3.head())

# Print the count of missing values for each column to verify imputation
print("\nMissing values count per column after imputation:")
print(f130to141_train_imputed_id2_id3[id2_grouped_cols].isnull().sum())

Streaming output truncated to the last 5000 lines.
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/usr/local/lib/python3.11/dist-packages/numpy/lib/_nanfunctions_impl.py:1231: RuntimeWarning: 

Head of f130to141_train_imputed_id2_id3 after imputation:


id2        id3      f130      f131      f132      f133      f134  \
0  1366776  189706075  0.012500  0.055556  0.000531  0.011765  0.015789   
1  1366776      89227  0.012500  0.055556  0.000531  0.011765  0.015789   
2  1366776      35046  0.015873  0.035088  0.000655  0.007092  0.013072   
3  1366776    6275451  0.012500  0.055556  0.000531  0.011765  0.015789   
4  1366776      78053  0.012500  0.055556  0.000531  0.011765  0.015789   

       f137      f138      f139      f140      f141  
0  0.004595  0.007035  1.531110  0.090909  0.363636  
1  0.004595  0.007035  1.531110  0.090909  0.363636  
2  0.003608  0.005545  1.536837  0.142857  0.285714  
3  0.004595  0.007035  1.531110  0.090909  0.363636  
4  0.004595  0.007035  1.531110  0.090909  0.363636


Missing values count per column after imputation:
f140    0
f141    0
dtype: int64


In [22]:
# Create a copy to avoid modifying the original dataframe
f130to141_test_imputed_id3 = f130to141_test_modified.copy()

# Define the columns to be imputed based on id3 (excluding f135 and f136 which were dropped)
id3_grouped_cols_test = [col for col in id3_grouped_cols if col in f130to141_test_imputed_id3.columns]

# Impute missing values in id3_grouped_cols_test
for col in id3_grouped_cols_test:
    # Calculate the global median for the current column
    global_median_test = f130to141_test_imputed_id3[col].median()

    # Group by 'id3' and fill missing values within each group with the group median
    # Use transform to return a Series with the same index as the original DataFrame
    f130to141_test_imputed_id3[col] = f130to141_test_imputed_id3.groupby('id3')[col].transform(
        lambda x: x.fillna(x.median())
    )

    # Fill any remaining missing values with the global median
    f130to141_test_imputed_id3[col] = f130to141_test_imputed_id3[col].fillna(global_median_test)

# Display the head and check for null values in the imputed columns
print("Head of f130to141_test_imputed_id3 after imputation:")
display(f130to141_test_imputed_id3.head())

print("\nChecking for null values in imputed columns (id3 grouped):")
print(f130to141_test_imputed_id3[id3_grouped_cols_test].isnull().sum())

Head of f130to141_test_imputed_id3 after imputation:


id2     id3      f130      f131      f132      f133      f134  \
46756   1362907   91950  0.000000  0.000000  0.000000  0.000000  0.000000   
57819   1082599   88356  0.022260  0.027778  0.027631  0.080645  0.031792   
15390   1888466  958700  0.333333  0.454545  0.425569  0.366667  0.463415   
145730  1888971  795739  0.000000  0.000000  0.000000  0.010283  0.000000   
146085  1256369   82296  0.000000  0.117647  0.010753  0.000000  0.034483   

            f137      f138      f139      f140      f141  
46756   0.000000  0.000000  1.037931       NaN       NaN  
57819   0.030366  0.038854  1.279486  0.074713  0.022989  
15390   0.423653  0.399015  0.941844  0.017668  0.017668  
145730  0.000000  0.006845  1.045526       NaN       NaN  
146085  0.013966  0.018868  1.350943  0.000000  0.200000


Checking for null values in imputed columns (id3 grouped):
f130    0
f131    0
f132    0
f133    0
f134    0
f137    0
f138    0
f139    0
dtype: int64


In [23]:
# Create a copy of the DataFrame
f130to141_test_imputed_id2_id3 = f130to141_test_imputed_id3.copy()

# Define the columns to be imputed in this subtask
id2_grouped_cols_test = ['f140', 'f141']

# Iterate through each column and perform grouped median imputation with global median fallback
for col in id2_grouped_cols_test:
    # Calculate the global median for the column
    global_median_test = f130to141_test_imputed_id2_id3[col].median()

    # Impute missing values by grouping by 'id2' and using the median of non-null values,
    # falling back to the global median if a group has no non-null values.
    f130to141_test_imputed_id2_id3[col] = f130to141_test_imputed_id2_id3.groupby('id2')[col].transform(
        lambda x: x.fillna(x.median() if pd.notna(x.median()) else global_median_test)
    )

# Display the head of the imputed DataFrame
print("Head of f130to141_test_imputed_id2_id3 after imputation:")
display(f130to141_test_imputed_id2_id3.head())

# Print the count of missing values for each column to verify imputation
print("\nMissing values count per column after imputation:")
print(f130to141_test_imputed_id2_id3[id2_grouped_cols_test].isnull().sum())

Head of f130to141_test_imputed_id2_id3 after imputation:


id2     id3      f130      f131      f132      f133      f134  \
46756   1362907   91950  0.000000  0.000000  0.000000  0.000000  0.000000   
57819   1082599   88356  0.022260  0.027778  0.027631  0.080645  0.031792   
15390   1888466  958700  0.333333  0.454545  0.425569  0.366667  0.463415   
145730  1888971  795739  0.000000  0.000000  0.000000  0.010283  0.000000   
146085  1256369   82296  0.000000  0.117647  0.010753  0.000000  0.034483   

            f137      f138      f139      f140      f141  
46756   0.000000  0.000000  1.037931  0.043478  0.014799  
57819   0.030366  0.038854  1.279486  0.074713  0.022989  
15390   0.423653  0.399015  0.941844  0.017668  0.017668  
145730  0.000000  0.006845  1.045526  0.043478  0.014799  
146085  0.013966  0.018868  1.350943  0.000000  0.200000


Missing values count per column after imputation:
f140    0
f141    0
dtype: int64


In [24]:
import numpy as np

# Rename dataframes
f130to141_train = f130to141_train_imputed_id2_id3
f130to141_test = f130to141_test_imputed_id2_id3

print("Dataframes renamed.")

# Analyze nulls and ranges (train)
print("\nAnalysis of f130to141_train after imputation:")
analysis_results_train_imputed = {}
for col in f130to141_train.columns:
    null_percentage = f130to141_train[col].isnull().mean() * 100
    col_range = None
    if pd.api.types.is_numeric_dtype(f130to141_train[col]):
        if f130to141_train[col].notna().any():
            col_range = f130to141_train[col].max() - f130to141_train[col].min()

    analysis_results_train_imputed[col] = {
        'null_percentage': null_percentage,
        'range': col_range
    }

for col, results in analysis_results_train_imputed.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['range'] is not None:
        print(f"  Range: {results['range']:.4f}")
    else:
        print("  Range: N/A")
    print("-" * 20)


# Analyze nulls and ranges (test)
print("\nAnalysis of f130to141_test after imputation:")
analysis_results_test_imputed = {}
for col in f130to141_test.columns:
    null_percentage = f130to141_test[col].isnull().mean() * 100
    col_range = None
    if pd.api.types.is_numeric_dtype(f130to141_test[col]):
        if f130to141_test[col].notna().any():
            col_range = f130to141_test[col].max() - f130to141_test[col].min()

    analysis_results_test_imputed[col] = {
        'null_percentage': null_percentage,
        'range': col_range
    }

for col, results in analysis_results_test_imputed.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['range'] is not None:
        print(f"  Range: {results['range']:.4f}")
    else:
        print("  Range: N/A")
    print("-" * 20)

Dataframes renamed.

Analysis of f130to141_train after imputation:
Column: id2
  Null Percentage: 0.00%
  Range: 910525.0000
--------------------
Column: id3
  Null Percentage: 0.00%
  Range: 876664650.0000
--------------------
Column: f130
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f131
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f132
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f133
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f134
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f137
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f138
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f139
  Null Percentage: 0.00%
  Range: 310.1153
--------------------
Column: f140
  Null Percentage: 0.00%
  Range: 1.0000
--------------------
Column: f141
  Null Percentage: 0.00%
  Range: 1.0000
--------------------

Analysis of f130to1

In [25]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler
import pandas as pd

# Ensure 'f139' is numeric in both dataframes, coercing errors
f130to141_train['f139'] = pd.to_numeric(f130to141_train['f139'], errors='coerce')
f130to141_test['f139'] = pd.to_numeric(f130to141_test['f139'], errors='coerce')


# Apply log1p transformation to 'f139'
f130to141_train['f139'] = np.log1p(f130to141_train['f139'])
f130to141_test['f139'] = np.log1p(f130to141_test['f139'])

# Initialize and fit MinMaxScaler on the training data
scaler = MinMaxScaler()
scaler.fit(f130to141_train[['f139']])

# Apply Min-Max scaling to 'f139' in both train and test dataframes
f130to141_train['f139'] = scaler.transform(f130to141_train[['f139']])
f130to141_test['f139'] = scaler.transform(f130to141_test[['f139']])

print("Transformations applied to 'f139' column.")

# Display the head of the modified dataframes and describe 'f139' to verify
print("\nHead of f130to141_train after transformations:")
display(f130to141_train.head())
print("\nDescription of 'f139' in f130to141_train after transformations:")
display(f130to141_train['f139'].describe())


print("\nHead of f130to141_test after transformations:")
display(f130to141_test.head())
print("\nDescription of 'f139' in f130to141_test after transformations:")
display(f130to141_test['f139'].describe())

Transformations applied to 'f139' column.

Head of f130to141_train after transformations:


id2        id3      f130      f131      f132      f133      f134  \
0  1366776  189706075  0.012500  0.055556  0.000531  0.011765  0.015789   
1  1366776      89227  0.012500  0.055556  0.000531  0.011765  0.015789   
2  1366776      35046  0.015873  0.035088  0.000655  0.007092  0.013072   
3  1366776    6275451  0.012500  0.055556  0.000531  0.011765  0.015789   
4  1366776      78053  0.012500  0.055556  0.000531  0.011765  0.015789   

       f137      f138      f139      f140      f141  
0  0.004595  0.007035  0.161782  0.090909  0.363636  
1  0.004595  0.007035  0.161782  0.090909  0.363636  
2  0.003608  0.005545  0.162176  0.142857  0.285714  
3  0.004595  0.007035  0.161782  0.090909  0.363636  
4  0.004595  0.007035  0.161782  0.090909  0.363636


Description of 'f139' in f130to141_train after transformations:


count    770164.000000
mean          0.120384
std           0.043287
min           0.000000
25%           0.118365
50%           0.123247
75%           0.131668
max           1.000000
Name: f139, dtype: float64


Head of f130to141_test after transformations:


id2     id3      f130      f131      f132      f133      f134  \
46756   1362907   91950  0.000000  0.000000  0.000000  0.000000  0.000000   
57819   1082599   88356  0.022260  0.027778  0.027631  0.080645  0.031792   
15390   1888466  958700  0.333333  0.454545  0.425569  0.366667  0.463415   
145730  1888971  795739  0.000000  0.000000  0.000000  0.010283  0.000000   
146085  1256369   82296  0.000000  0.117647  0.010753  0.000000  0.034483   

            f137      f138      f139      f140      f141  
46756   0.000000  0.000000  0.124027  0.043478  0.014799  
57819   0.030366  0.038854  0.143541  0.074713  0.022989  
15390   0.423653  0.399015  0.115613  0.017668  0.017668  
145730  0.000000  0.006845  0.124675  0.043478  0.014799  
146085  0.013966  0.018868  0.148919  0.000000  0.200000


Description of 'f139' in f130to141_test after transformations:


count    369301.000000
mean          0.121159
std           0.048716
min           0.000000
25%           0.118618
50%           0.124095
75%           0.133159
max           0.994242
Name: f139, dtype: float64

In [26]:
import numpy as np
import pandas as pd

# Analyze nulls and ranges (train) for the final dataframe
print("Final Analysis of f130to141_train:")
final_analysis_results_train = {}
for col in f130to141_train.columns:
    null_percentage = f130to141_train[col].isnull().mean() * 100
    min_val = None
    max_val = None
    if pd.api.types.is_numeric_dtype(f130to141_train[col]):
        if f130to141_train[col].notna().any():
            min_val = f130to141_train[col].min()
            max_val = f130to141_train[col].max()

    final_analysis_results_train[col] = {
        'null_percentage': null_percentage,
        'min': min_val,
        'max': max_val
    }

for col, results in final_analysis_results_train.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['min'] is not None and results['max'] is not None:
        print(f"  Range: Min = {results['min']:.4f}, Max = {results['max']:.4f}")
    else:
        print("  Range: N/A (column is not numeric or all values are null)")
    print("-" * 20)


# Analyze nulls and ranges (test) for the final dataframe
print("\nFinal Analysis of f130to141_test:")
final_analysis_results_test = {}
for col in f130to141_test.columns:
    null_percentage = f130to141_test[col].isnull().mean() * 100
    min_val = None
    max_val = None
    if pd.api.types.is_numeric_dtype(f130to141_test[col]):
        if f130to141_test[col].notna().any():
            min_val = f130to141_test[col].min()
            max_val = f130to141_test[col].max()

    final_analysis_results_test[col] = {
        'null_percentage': null_percentage,
        'min': min_val,
        'max': max_val
    }

for col, results in final_analysis_results_test.items():
    print(f"Column: {col}")
    print(f"  Null Percentage: {results['null_percentage']:.2f}%")
    if results['min'] is not None and results['max'] is not None:
        print(f"  Range: Min = {results['min']:.4f}, Max = {results['max']:.4f}")
    else:
        print("  Range: N/A (column is not numeric or all values are null)")
    print("-" * 20)

Final Analysis of f130to141_train:
Column: id2
  Null Percentage: 0.00%
  Range: Min = 1000043.0000, Max = 1910568.0000
--------------------
Column: id3
  Null Percentage: 0.00%
  Range: Min = 1185.0000, Max = 876665835.0000
--------------------
Column: f130
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f131
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f132
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f133
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f134
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f137
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f138
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
--------------------
Column: f139
  Null Percentage: 0.00%
  Range: Min = 0.0000, Max = 1.0000
---------------

In [27]:
# Drop 'id2' from f142to309_train and show shapes
print("Shape of f142to309_train before dropping 'id2':", f142to309_train.shape)
f142to309_train_modified = f142to309_train.drop('id2', axis=1)
print("Shape of f142to309_train after dropping 'id2':", f142to309_train_modified.shape)

# Drop 'id2' from f142to309_test and show shapes
print("\nShape of f142to309_test before dropping 'id2':", f142to309_test.shape)
f142to309_test_modified = f142to309_test.drop('id2', axis=1)
print("Shape of f142to309_test after dropping 'id2':", f142to309_test_modified.shape)

# Drop 'id3' from f130to141_train and show shapes
print("\nShape of f130to141_train before dropping 'id3':", f130to141_train.shape)
f130to141_train_modified = f130to141_train.drop('id3', axis=1)
print("Shape of f130to141_train after dropping 'id3':", f130to141_train_modified.shape)

# Drop 'id3' from f130to141_test and show shapes
print("\nShape of f130to141_test before dropping 'id3':", f130to141_test.shape)
f130to141_test_modified = f130to141_test.drop('id3', axis=1)
print("Shape of f130to141_test after dropping 'id3':", f130to141_test_modified.shape)

# Merge train dataframes (f130to141 on left) and name the result f130to309_train
f130to309_train = pd.merge(f130to141_train_modified, f142to309_train_modified, left_index=True, right_index=True)
print("\nShape of merged train dataframe f130to309_train:", f130to309_train.shape)

# Merge test dataframes (f130to141 on left) and name the result f130to309_test
f130to309_test = pd.merge(f130to141_test_modified, f142to309_test_modified, left_index=True, right_index=True)
print("Shape of merged test dataframe f130to309_test:", f130to309_test.shape)

# Verify final dataframes
print("\nHead of final train dataframe f130to309_train:")
display(f130to309_train.head())

print("\nHead of final test dataframe f130to309_test:")
display(f130to309_test.head())

Shape of f142to309_train before dropping 'id2': (770164, 168)
Shape of f142to309_train after dropping 'id2': (770164, 167)

Shape of f142to309_test before dropping 'id2': (369301, 168)
Shape of f142to309_test after dropping 'id2': (369301, 167)

Shape of f130to141_train before dropping 'id3': (770164, 12)
Shape of f130to141_train after dropping 'id3': (770164, 11)

Shape of f130to141_test before dropping 'id3': (369301, 12)
Shape of f130to141_test after dropping 'id3': (369301, 11)

Shape of merged train dataframe f130to309_train: (770164, 178)
Shape of merged test dataframe f130to309_test: (369301, 178)

Head of final train dataframe f130to309_train:


id2      f130      f131      f132      f133      f134      f137  \
0  1366776  0.012500  0.055556  0.000531  0.011765  0.015789  0.004595   
1  1366776  0.012500  0.055556  0.000531  0.011765  0.015789  0.004595   
2  1366776  0.015873  0.035088  0.000655  0.007092  0.013072  0.003608   
3  1366776  0.012500  0.055556  0.000531  0.011765  0.015789  0.004595   
4  1366776  0.012500  0.055556  0.000531  0.011765  0.015789  0.004595   

       f138      f139      f140  ...  f301  f302  f303  f304  f305  f306  \
0  0.007035  0.161782  0.090909  ...     0     0     0     0     0     0   
1  0.007035  0.161782  0.090909  ...     0     0     0     0     0     0   
2  0.005545  0.162176  0.142857  ...     0     0     0     0     0     0   
3  0.007035  0.161782  0.090909  ...     0     0     0     0     0     0   
4  0.007035  0.161782  0.090909  ...     0     0     0     0     0     0   

   f307  f308  f309  missing_offer_meta  
0     0     0     0                   0  
1     0     0     0                   0  
2     0     0     0                   0  
3     0     0     0                   0  
4     0     0     0                   0  

[5 rows x 178 columns]


Head of final test dataframe f130to309_test:


id2      f130      f131      f132      f133      f134      f137  \
46756   1362907  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
57819   1082599  0.022260  0.027778  0.027631  0.080645  0.031792  0.030366   
15390   1888466  0.333333  0.454545  0.425569  0.366667  0.463415  0.423653   
145730  1888971  0.000000  0.000000  0.000000  0.010283  0.000000  0.000000   
146085  1256369  0.000000  0.117647  0.010753  0.000000  0.034483  0.013966   

            f138      f139      f140  ...  f301  f302  f303  f304  f305  f306  \
46756   0.000000  0.124027  0.043478  ...     0     0     0     0     0     0   
57819   0.038854  0.143541  0.074713  ...     0     0     0     0     0     0   
15390   0.399015  0.115613  0.017668  ...     0     0     0     0     0     0   
145730  0.006845  0.124675  0.043478  ...     0     0     0     0     0     0   
146085  0.018868  0.148919  0.000000  ...     0     0     0     0     0     0   

        f307  f308  f309  missing_offer_meta  
46756      0     0     0                   0  
57819      0     0     0                   0  
15390      0     0     0                   0  
145730     0     0     0                   0  
146085     0     0     0                   0  

[5 rows x 178 columns]

In [28]:
output_path_train = '/content/drive/MyDrive/Transformed_Train_Data/f130to309_train.parquet'
f130to309_train.to_parquet(output_path_train, index=False)

print(f"DataFrame f130to309_train saved to {output_path_train}")

output_path_test = '/content/drive/MyDrive/Transformed_Test_Data/f130to309_test.parquet'
f130to309_test.to_parquet(output_path_test, index=False)

print(f"DataFrame f130to309_test saved to {output_path_test}")

DataFrame f130to309_train saved to /content/drive/MyDrive/Transformed_Train_Data/f130to309_train.parquet
DataFrame f130to309_test saved to /content/drive/MyDrive/Transformed_Test_Data/f130to309_test.parquet
